### PR TITLE
Caw function and testing [WIP]

### PR DESCRIPTION
**Pull request by wsuppiger**
Here are the main changes for this PR:
- modified caw function class to be static and include all caw calls (some are WIP)
- modified faz server to add in if else statements and change to enum event types
- modified kv store and client to include an interface (abstract class) for Dependency Injection. Credit Adam for the idea.
- added testing for caw function (for Registeruser, Follow, and Profile)

### CODE REVIEW

File: src/key_value/key_value_interface.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#define SRC_KEY_VALUE_KEY_VALUE_INTERFACE_H_
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace csci499 {
+
+// key value storage for csci499 application
+class KeyValueInterface {
```
Comment (barath) [inline src/key_value/key_value_interface.h:14]: Could you add more descriptive comments for the class and interfaces?  Especially explain why the interface exists.

File: src/key_value/key_value_client.h [hunk 2]
```diff
@@ -11,29 +11,29 @@
...
+class KeyValueClient : public KeyValueInterface {
  public:
   explicit KeyValueClient(std::shared_ptr<Channel> channel)
       : stub_(KeyValueStore::NewStub(channel)) {}
 
   virtual ~KeyValueClient() {}
 
   // package and send rpc call put to server
-  bool Put(const std::string& key, const std::string& value);
+  virtual void Put(const std::string& key, const std::string& value);
```
Comment (barath) [inline src/key_value/key_value_client.h]: Use the `override` keyword where appropriate

Comment (barath) [inline src/key_value/key_value_client.h]: If this implements the interface exactly, you don't need method comments for any method that overrides the parent for whom the parent's method comment is descriptive enough.

File: src/key_value/key_value.h [hunk 3]
```diff
@@ -9,24 +9,25 @@
 #include <unordered_map>
 #include <vector>
 
+#include "key_value_interface.h"
+
 namespace csci499 {
 
 // key value storage for csci499 application
-class KeyValue {
+class KeyValue : public KeyValueInterface {
```
Comment (barath) [inline src/key_value/key_value.h:17]: Same comment from below applies here to method comments and override

File: src/faz/faz_server.cc [hunk 4]
```diff
@@ -9,6 +9,13 @@
 #include <vector>
 
 namespace csci499 {
+enum EventType {
```
Comment (barath) [inline src/faz/faz_server.cc:13]: Add comments

File: src/caw/caw_function.cc [hunk 5]
```diff
@@ -3,17 +3,102 @@
...
+  payload.UnpackTo(&request);
+  std::string username = request.username();
+  if (!UserExists(username, kv)) {
+    LOG(WARNING) << "the user does not exist";
+    return {Status(StatusCode::UNAVAILABLE, "the user does not exist")};
+  }
+  std::vector<std::string> followers = kv.Get("ufollowers-" + username);
+  std::vector<std::string> following = kv.Get("ufollowing-" + username);
+  ProfileReply reply;
+  for (auto f : followers) {
```
Comment (barath) [inline src/caw/caw_function.cc]: Prefer const auto &

Comment (wsuppiger) [reply src/caw/caw_function.cc]: fixing in next PR

Comment (barath) [reply src/caw/caw_function.cc]: Just fyi, it's a best practice to fix issues like this in the PR that they're raised in, so the logs are easy to follow later.  That is, if you get a comment on something in one PR and plan to fix it, you should either add a TODO comment or make the fix in this PR.  It's small enough that it's better to just make the edit.

Comment (wsuppiger) [reply src/caw/caw_function.cc]: Sounds good, I have just been having trouble with branch management, so if I commit caw_function.cc there will be more changes than just `const auto &`.  I think I am starting to figure it out, but I have had a few hiccups.

**Pull request merged**